genie-web/src/main/java/com/netflix/genie/web/rpc/grpc/services/impl/v4/GRpcJobTerminationServiceImpl.java

```diff
@@ -0,0 +1,91 @@
+ *     you may not use this file except in compliance with the License.
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+package com.netflix.genie.web.rpc.grpc.services.impl.v4;
+
+import com.netflix.genie.proto.JobKillListenerRequest;
+import com.netflix.genie.proto.JobKillListenerResponse;
+import com.netflix.genie.proto.JobTerminationServiceGrpc;
+import com.netflix.genie.web.rpc.grpc.interceptors.SimpleLoggingInterceptor;
+import com.netflix.genie.web.services.JobKillService;
+import io.grpc.stub.StreamObserver;
+import lombok.extern.slf4j.Slf4j;
+import net.devh.springboot.autoconfigure.grpc.server.GrpcService;
+
+import java.util.function.Consumer;
+
+/**
+ * Extension of {@link JobTerminationServiceGrpc.JobTerminationServiceImplBase}
+ * to provide functionality for killing jobs running on agents.
+ *
+ * @author standon
+ * @since 4.0.0
+ */
+@GrpcService(
+    value = JobTerminationServiceGrpc.class,
+    interceptors = {
+        SimpleLoggingInterceptor.class,
+    }
+)
+@Slf4j
+public class GRpcJobTerminationServiceImpl extends JobTerminationServiceGrpc.JobTerminationServiceImplBase {
+
+    private final JobKillService jobKillService;
+
+    /**
+     * Constructor.
+     *
+     * @param jobKillService The implementation of the {@link JobKillService} to use
+     */
+    public GRpcJobTerminationServiceImpl(final JobKillService jobKillService) {
+        this.jobKillService = jobKillService;
+    }
+
+    /**
+     * Register to get a callback when a job is killed.
+     *
+     * @param request          Request to listen for job registerCallback
+     * @param responseObserver The response observer
+     */
+    @Override
+    public void registerForKillNotification(
+        final JobKillListenerRequest request,
+        final StreamObserver<JobKillListenerResponse> responseObserver
+    ) {
+        jobKillService.registerCallbackForKill(
+            request.getJobId(),
+            new JobKillConsumer(responseObserver)
+        );
+    }
+
+    private static class JobKillConsumer implements Consumer<String> {
+
+        private StreamObserver<JobKillListenerResponse> responseObserver;
+
+        JobKillConsumer(final StreamObserver<JobKillListenerResponse> responseObserver) {
+            this.responseObserver = responseObserver;
+        }
+
+        @Override
+        public void accept(final String id) {
+            responseObserver.onNext(
+                JobKillListenerResponse.newBuilder().setJobId(id).build()
+            );
+            responseObserver.onCompleted();
```

Comment:
Once you've called `onCompleted` on the observer you can't use this `responseObserver` anymore. It will throw exceptions as you try to use it. So with that in mind once this i called this listener should be destroyed.

Reply:
Done